drivers/hmis/app/models/hmis/hud/enrollment.rb

```diff
@@ -51,7 +51,9 @@ class Hmis::Hud::Enrollment < Hmis::Hud::Base
   # hide previous declaration of :viewable_by, we'll use this one
   # A user can see any enrollment associated with a project they can access
   replace_scope :viewable_by, ->(user) do
-    project_ids = Hmis::Hud::Project.viewable_by(user).pluck(:id, :ProjectID)
+    project_ids = Hmis::Hud::Project.viewable_by(user).
+      select { |project| user.can_view_enrollment_details_for?(project) }.
+      pluck(:id, :ProjectID)
```

Comment:
I think this `select` addition is causing an infinite loop? Or, just taking a really long time. I'm consistently getting 504s on the View Project page for the Enrollments table (and it only has 6 enrollments) and the "All Enrollments" page on the client dashboard.

Can you repro that? If I comment out line 55, it speeds up considerably.

For the Project case – resolving all enrollments on a project – we should really only need to call `user.can_view_enrollment_details_for?(project)` once.

Reply:
I don't know exactly what part is so slow. But it may be quicker to say "give me all the project IDs that I `can_view_enrollment_details` for" and filter down the enrollment scope by that.

```
projects = Hmis::Hud::Project.with_access(user, :can_view_enrollment_details)
scope.viewable_by(user).in_project_including_wip(ids, project_ids)
```

If we do the permission check outside the `viewable_by` scope, we can bypass it entirely for the case when we are resolving enrollments on 1 project only. (In that case we just need to ask `user.can_view_enrollment_details_for?(project)` once and be done with it).

Reply:
Since `viewable_by` returns a scope, and you can call `Fu.scope.select(:id, :created_at).second_scope` it's probably confused.  I _think_ you want:
```suggestion
    project_ids = Hmis::Hud::Project.viewable_by(user).to_a.
      select { |project| user.can_view_enrollment_details_for?(project) }.
      pluck(:id, :ProjectID)
```

Reply:
That doesn't scale particularly well, is there any way to add a scope to project that could join your access lists where the permission is appropriately set?

Reply:
How about:
```suggestion
    project_ids = Hmis::Hud::Project.with_access(user, :can_view_enrollment_details).viewable_by(user).
      pluck(:id, :ProjectID)
```

Reply:
That is pretty much exactly what I suggested, no?

Forget what I said about moving this out of the `viewable_by` scope, though. I think the `can_view_enrollment_details` permission check should continue to happen in `viewable_by`. Even though it will sometimes be redundant, it'll be safer.

Reply:
Might be the same, I didn't think 
```
projects = Hmis::Hud::Project.with_access(user, :can_view_enrollment_details)
scope.viewable_by(user).in_project_including_wip(ids, project_ids)
```
looked like it would be runnable, and I wasn't following what was a scope and what was instantiated.

Reply:
Ok, yeah,well I agree with suggestion from https://github.com/greenriver/hmis-warehouse/pull/2868#discussion_r1135952383, the `with_access` scope just needs to be created and not be terribly slow 🙂